### PR TITLE
TMA-662: Users Brick passes with empty input source

### PR DESCRIPTION
**Pull request by panjan**

### CODE REVIEW

File: spec/unit/actions/synchronize_users_spec.rb [hunk 1]
```diff
@@ -44,8 +48,25 @@
         GoodData::LCM2.convert_to_smart_hash(params)
       end
 
-      it_behaves_like 'a user action reading client_id' do
-        let(:client_id) { '123456789' }
+      context 'when the input set is empty' do
```
Comment (yenkins) [inline spec/unit/actions/synchronize_users_spec.rb]: Similar code found in synchronize_users_spec.rb:51, synchronize_users_spec.rb:61 (mass = 64)

File: spec/unit/actions/synchronize_users_spec.rb [hunk 2]
```diff
@@ -44,8 +48,25 @@
...
+        before do
+          allow(File).to receive(:open).and_return("")
+        end
+        it 'does not fail and logs a warning' do
+          expect(logger).to receive(:warn).with(/not match any client ids/)
+          expect { subject.class.call(params) }.to_not raise_error
+        end
+      end
+
+      context 'when the input set does not contain data for the current project' do
```
Comment (yenkins) [inline spec/unit/actions/synchronize_users_spec.rb]: Similar code found in synchronize_users_spec.rb:51, synchronize_users_spec.rb:61 (mass = 64)

File: lib/gooddata/lcm/actions/synchronize_users.rb [hunk 3]
```diff
@@ -125,18 +126,21 @@ def call(params)
...
                         fail "Column for determining the project assignement is empty for \"#{u[:login]}\"" if u[:pid].blank?
                         client_id = u[:pid].to_s
-                        (goodot_id && client_id == goodot_id) || domain.clients(client_id, data_product).project_uri == project.uri
+                        client_id == filter_value
                       end
 
                       if filtered_users.empty?
-                        fail "Project \"#{project.pid}\" does not match with any client ids in input source (both GOODOT_CUSTOM_PROJECT_ID and SEGMENT/CLIENT). \
-  We are unable to get the value to filter users."
+                        params.gdc_logger.warn(
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_users.rb]: Calls 'params.gdc_logger' 2 times (DuplicateMethodCall)

File: lib/gooddata/lcm/actions/synchronize_users.rb [hunk 4]
```diff
@@ -125,18 +126,21 @@ def call(params)
...
+                          "Project \"#{project.pid}\" does not match " \
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_users.rb]: Calls 'project.pid' 7 times (DuplicateMethodCall)

File: lib/gooddata/lcm/user_bricks_helper.rb [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+          md = project.metadata
+          goodot_id = md['GOODOT_CUSTOM_PROJECT_ID'].to_s
+
+          client = domain.clients.find { |c| c.project_uri == project.uri }
+          if goodot_id.empty? && client.nil?
+            fail "Project \"#{project.pid}\" metadata does not contain key GOODOT_CUSTOM_PROJECT_ID neither is it mapped \
+            to a client_id in LCM metadata. We are unable to get the values for user filters."
+          end
+
+          unless goodot_id.empty? || client.nil? || (goodot_id == client.id)
```
Comment (yenkins) [inline lib/gooddata/lcm/user_bricks_helper.rb]: Calls 'client.id' 2 times (DuplicateMethodCall)

File: lib/gooddata/lcm/user_bricks_helper.rb [hunk 6]
```diff
@@ -0,0 +1,25 @@
+module GoodData
+  module LCM2
+    class UserBricksHelper
+      class << self
+        def resolve_client_id(domain, project)
+          md = project.metadata
+          goodot_id = md['GOODOT_CUSTOM_PROJECT_ID'].to_s
+
+          client = domain.clients.find { |c| c.project_uri == project.uri }
+          if goodot_id.empty? && client.nil?
```
Comment (yenkins) [inline lib/gooddata/lcm/user_bricks_helper.rb]: Calls 'client.nil?' 2 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/user_bricks_helper.rb]: Calls 'goodot_id.empty?' 3 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/user_bricks_helper.rb]: Refers to 'client' more than self (maybe move it to another class?) (FeatureEnvy)

Comment (yenkins) [inline lib/gooddata/lcm/user_bricks_helper.rb]: Refers to 'goodot_id' more than self (maybe move it to another class?) (FeatureEnvy)

Comment (yenkins) [inline lib/gooddata/lcm/user_bricks_helper.rb]: Performs a nil-check (NilCheck)

File: lib/gooddata/lcm/user_bricks_helper.rb [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+            fail "Project \"#{project.pid}\" metadata does not contain key GOODOT_CUSTOM_PROJECT_ID neither is it mapped \
```
Comment (yenkins) [inline lib/gooddata/lcm/user_bricks_helper.rb]: Calls 'project.pid' 2 times (DuplicateMethodCall)

File: lib/gooddata/lcm/user_bricks_helper.rb [hunk 8]
```diff
@@ -0,0 +1,25 @@
+module GoodData
+  module LCM2
+    class UserBricksHelper
+      class << self
+        def resolve_client_id(domain, project)
+          md = project.metadata
```
Comment (yenkins) [inline lib/gooddata/lcm/user_bricks_helper.rb]: Refers to 'project' more than self (maybe move it to another class?) (FeatureEnvy)

File: lib/gooddata/lcm/user_bricks_helper.rb [hunk 9]
```diff
@@ -0,0 +1,25 @@
+module GoodData
+  module LCM2
+    class UserBricksHelper
```
Comment (yenkins) [inline lib/gooddata/lcm/user_bricks_helper.rb]: Has no descriptive comment (IrresponsibleModule)

File: lib/gooddata/lcm/user_bricks_helper.rb [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+      class << self
+        def resolve_client_id(domain, project)
```
Comment (yenkins) [inline lib/gooddata/lcm/user_bricks_helper.rb]: Has approx 8 statements (TooManyStatements)

File: lib/gooddata/lcm/user_bricks_helper.rb [hunk 11]
```diff
@@ -0,0 +1,25 @@
+module GoodData
+  module LCM2
```
Comment (yenkins) [inline lib/gooddata/lcm/user_bricks_helper.rb]: Has the name 'lcm2' (UncommunicativeModuleName)

File: lib/gooddata/lcm/user_bricks_helper.rb [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+    class UserBricksHelper
+      class << self
+        def resolve_client_id(domain, project)
+          md = project.metadata
+          goodot_id = md['GOODOT_CUSTOM_PROJECT_ID'].to_s
+
+          client = domain.clients.find { |c| c.project_uri == project.uri }
```
Comment (yenkins) [inline lib/gooddata/lcm/user_bricks_helper.rb]: Has the variable name 'c' (UncommunicativeVariableName)

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/598/ : SUCCESS in 12s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/403/ : FAILURE in 2m 08s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/711/ : SUCCESS in 2m 01s

**Comment by panjan** [comment]
extended test

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/599/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-ruby-integration-tests-zuul/1140/ : FAILURE in 48m 24s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/898/ : UNSTABLE in 30m 33s

File: lib/gooddata/lcm/user_bricks_helper.rb [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+    class UserBricksHelper
+      class << self
+        def resolve_client_id(domain, project, data_product)
+          md = project.metadata
+          goodot_id = md['GOODOT_CUSTOM_PROJECT_ID'].to_s
+
+          client = domain.clients(:all, data_product).find do |c|
+            c.project_uri == project.uri
+          end
+          if goodot_id.empty? && client.nil?
```
Comment (yenkins) [inline lib/gooddata/lcm/user_bricks_helper.rb]: Calls 'client.nil?' 2 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/user_bricks_helper.rb]: Calls 'goodot_id.empty?' 3 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/user_bricks_helper.rb]: Refers to 'client' more than self (maybe move it to another class?) (FeatureEnvy)

Comment (yenkins) [inline lib/gooddata/lcm/user_bricks_helper.rb]: Refers to 'goodot_id' more than self (maybe move it to another class?) (FeatureEnvy)

Comment (yenkins) [inline lib/gooddata/lcm/user_bricks_helper.rb]: Performs a nil-check (NilCheck)

File: lib/gooddata/lcm/user_bricks_helper.rb [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+            fail "Project \"#{project.pid}\" metadata does not contain key GOODOT_CUSTOM_PROJECT_ID neither is it mapped \
```
Comment (yenkins) [inline lib/gooddata/lcm/user_bricks_helper.rb]: Calls 'project.pid' 2 times (DuplicateMethodCall)

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/600/ : SUCCESS in 16s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/404/ : FAILURE in 2m 07s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/712/ : SUCCESS in 2m 10s

**Comment by panjan** [comment]
extended test - staging

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/601/ : SUCCESS in 9s
- https://ci.intgdc.com/job/gooddata-ruby-integration-tests-zuul/1141/ : FAILURE in 44m 55s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/899/ : UNSTABLE in 2m 43s

**Comment by panjan** [comment]
extended test

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/602/ : SUCCESS in 7s
- https://ci.intgdc.com/job/gooddata-ruby-integration-tests-zuul/1142/ : SUCCESS in 49m 14s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/900/ : SUCCESS in 31m 27s

File: lib/gooddata/lcm/user_bricks_helper.rb [hunk 15]
```diff
@@ -0,0 +1,27 @@
+module GoodData
```
Comment (kubamahnert) [inline lib/gooddata/lcm/user_bricks_helper.rb]: Why no user_bricks_helper_spec ?

File: spec/unit/lcm/user_bricks_helper_spec.rb [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+  describe '.resolve_client_id' do
+    context 'when GOODOT_CUSTOM_PROJECT_ID specified' do
+      let(:metadata) { { 'GOODOT_CUSTOM_PROJECT_ID' => client_id } }
+      context 'when client matches specified project url' do
+        it 'returns client id' do
+          expect(subject).to eq(client_id)
+        end
+      end
+
+      context 'when goodot id does not match client id' do
```
Comment (yenkins) [inline spec/unit/lcm/user_bricks_helper_spec.rb]: Similar code found in user_bricks_helper_spec.rb:31, user_bricks_helper_spec.rb:48 (mass = 50)

File: spec/unit/lcm/user_bricks_helper_spec.rb [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+
+    context 'when GOODOT_CUSTOM_PROJECT_ID not specified' do
+      let(:metadata) { { 'GOODOT_CUSTOM_PROJECT_ID' => nil } }
+      context 'when client matches specified project url' do
+        it 'returns client id' do
+          expect(subject).to eq(client_id)
+        end
+      end
+
+      context 'when no client matches project url' do
```
Comment (yenkins) [inline spec/unit/lcm/user_bricks_helper_spec.rb]: Similar code found in user_bricks_helper_spec.rb:31, user_bricks_helper_spec.rb:48 (mass = 50)

File: lib/gooddata/lcm/user_bricks_helper.rb [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+            c.project_uri == project.uri
+          end
+          if goodot_id.empty? && client.nil?
+            fail "Project \"#{project.pid}\" metadata does not contain key " \
+                 "GOODOT_CUSTOM_PROJECT_ID neither is it mapped to a " \
+                 "client_id in LCM metadata. We are unable to get the " \
+                 "values for user filters."
+          end
+
+          unless goodot_id.empty? || client.nil? || (goodot_id == client.id)
```
Comment (yenkins) [inline lib/gooddata/lcm/user_bricks_helper.rb]: Calls 'client.id' 2 times (DuplicateMethodCall)

File: lib/gooddata/lcm/user_bricks_helper.rb [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+      class << self
+        def resolve_client_id(domain, project, data_product)
+          md = project.metadata
+          goodot_id = md['GOODOT_CUSTOM_PROJECT_ID'].to_s
+
+          client = domain.clients(:all, data_product).find do |c|
+            c.project_uri == project.uri
+          end
+          if goodot_id.empty? && client.nil?
+            fail "Project \"#{project.pid}\" metadata does not contain key " \
```
Comment (yenkins) [inline lib/gooddata/lcm/user_bricks_helper.rb]: Calls 'project.pid' 2 times (DuplicateMethodCall)

File: lib/gooddata/lcm/user_bricks_helper.rb [hunk 20]
```diff
@@ -0,0 +1,32 @@
+module GoodData
+  module LCM2
+    # Contains code shared by Users Brick and User Filters Brick
+    class UserBricksHelper
+      class << self
+        def resolve_client_id(domain, project, data_product)
+          md = project.metadata
```
Comment (yenkins) [inline lib/gooddata/lcm/user_bricks_helper.rb]: Refers to 'project' more than self (maybe move it to another class?) (FeatureEnvy)

File: lib/gooddata/lcm/user_bricks_helper.rb [hunk 21]
```diff
@@ -0,0 +1,32 @@
+module GoodData
+  module LCM2
+    # Contains code shared by Users Brick and User Filters Brick
+    class UserBricksHelper
+      class << self
+        def resolve_client_id(domain, project, data_product)
```
Comment (yenkins) [inline lib/gooddata/lcm/user_bricks_helper.rb]: Has approx 8 statements (TooManyStatements)

File: lib/gooddata/lcm/user_bricks_helper.rb [hunk 22]
```diff
@@ -0,0 +1,32 @@
...
+          md = project.metadata
+          goodot_id = md['GOODOT_CUSTOM_PROJECT_ID'].to_s
+
+          client = domain.clients(:all, data_product).find do |c|
```
Comment (yenkins) [inline lib/gooddata/lcm/user_bricks_helper.rb]: Has the variable name 'c' (UncommunicativeVariableName)

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/603/ : SUCCESS in 15s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/405/ : FAILURE in 2m 25s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/713/ : SUCCESS in 2m 05s

File: spec/unit/lcm/user_bricks_helper_spec.rb [hunk 23]
```diff
@@ -0,0 +1,59 @@
...
+      end
+    end
+  end
+
+  describe '.resolve_client_id' do
+    context 'when GOODOT_CUSTOM_PROJECT_ID specified' do
+      let(:metadata) { { 'GOODOT_CUSTOM_PROJECT_ID' => client_id } }
+      it_behaves_like 'a client id resolver'
+
+      context 'when goodot id does not match client id' do
```
Comment (yenkins) [inline spec/unit/lcm/user_bricks_helper_spec.rb]: Similar code found in user_bricks_helper_spec.rb:34, user_bricks_helper_spec.rb:47 (mass = 50)

File: spec/unit/lcm/user_bricks_helper_spec.rb [hunk 24]
```diff
@@ -0,0 +1,59 @@
...
+          expect { subject }.to raise_error(expected)
+        end
+      end
+    end
+
+    context 'when GOODOT_CUSTOM_PROJECT_ID not specified' do
+      let(:metadata) { { 'GOODOT_CUSTOM_PROJECT_ID' => nil } }
+      it_behaves_like 'a client id resolver'
+
+      context 'when no client matches project url' do
```
Comment (yenkins) [inline spec/unit/lcm/user_bricks_helper_spec.rb]: Similar code found in user_bricks_helper_spec.rb:34, user_bricks_helper_spec.rb:47 (mass = 50)

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/604/ : SUCCESS in 9s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/406/ : FAILURE in 34s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/714/ : SUCCESS in 32s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/605/ : SUCCESS in 12s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/407/ : SUCCESS in 30s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/715/ : SUCCESS in 32s

**Comment by panjan** [comment]
extended test

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/606/ : SUCCESS in 7s
- https://ci.intgdc.com/job/gooddata-ruby-integration-tests-zuul/1143/ : FAILURE in 48m 54s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/901/ : UNSTABLE in 13m 43s

**Comment by panjan** [comment]
extended test

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/609/ : SUCCESS in 7s
- https://ci.intgdc.com/job/gooddata-ruby-integration-tests-zuul/1145/ : FAILURE in 1h 15m 49s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/903/ : SUCCESS in 28m 03s

**Comment by panjan** [comment]
extended test

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/610/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-ruby-integration-tests-zuul/1146/ : SUCCESS in 49m 12s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/904/ : SUCCESS in 30m 55s

**Pull request merged**